packages/server/src/rooms.ts

```diff
@@ -1,48 +1,53 @@
 import { Server as IOServer, Socket } from 'socket.io';
-import { RoomEvent } from 'shared';
 
-import { Store } from './Store';
+import { Store, Room, Player } from './Store';
+import { RoomEvent } from 'shared';
 import { generateRoomName } from './utils';
 
 const MAX_ROOM_SIZE = 2;
 
-export async function createRoom(io: IOServer, socket: Socket, store: Store) {
+export function createRoom(io: IOServer, socket: Socket, store: Store, username: string) {
   const roomName = generateRoomName();
 
-  // @ts-expect-error
-  // TS2445: Property 'rooms' is protected and only accessible within class 'Adapter' and its subclasses.
-  const room = await io.sockets.adapter.rooms.get(roomName);
-  if (room === undefined || room.size < MAX_ROOM_SIZE) {
-    await socket.join(roomName);
-    store.rooms.push(roomName);
-
-    // TODO: Use a common logging format
-    console.info(`[CREATE] Client created and joined room ${roomName}`);
-    socket.emit(RoomEvent.CreateSuccess, roomName);
-  } else {
-    console.warn(`[CREATE FAILED] Client denied join, as room ${roomName} is full`);
-    socket.emit(RoomEvent.CreateError, 'Room is full');
+  const room = store.getRoom(roomName);
+  if (room !== undefined) {
+    console.warn(`[CREATE FAILED] Room ${roomName} already exists`);
+    socket.emit(RoomEvent.CreateError, 'Room already exists');
+    return;
   }
+
+  const newRoom = new Room(roomName);
+  store.addRoom(newRoom);
+  console.info(`[CREATE] Created room ${roomName}`);
+  socket.emit(RoomEvent.CreateSuccess, roomName);
 }
 
-export async function joinRoom(io: IOServer, socket: Socket, store: Store, roomName: string) {
-  // @ts-expect-error
-  const room = await io.sockets.adapter.rooms.get(roomName);
+export async function joinRoom(io: IOServer, socket: Socket, store: Store, roomName: string, username: string) {
+  const room = store.getRoom(roomName);
+
   if (room === undefined) {
-    if (store.rooms.includes(roomName)) {
-      await socket.join(roomName);
-      console.info(`[JOIN] Client joined room ${roomName}`);
-      socket.emit(RoomEvent.JoinSuccess);
-    } else {
-      console.warn(`[JOIN FAILED] Room ${roomName} does not exist`);
-      socket.emit(RoomEvent.JoinError, 'Room does not exist');
-    }
-  } else if (room.size >= MAX_ROOM_SIZE) {
+    console.warn(`[JOIN FAILED] Room ${roomName} does not exist`);
+    socket.emit(RoomEvent.JoinError, 'Room does not exist');
+    return;
+  }
+
+  const player = room.players.find((player) => player.name === username);
```

Comment:
this won't be expensive because we will have 2 / 4 players max but should we make a map (object) instead of a plain array using the player id or username as a key to avoid the need to use find?

Reply:
I expect this to be reviewed later and change a lot anyway... I think a relational model might be more suited, where you have room, player, and many to many relationships between them. Also even if it stays as an array, the `find` could be part of the `store`/`Room` and is not exposed like that. I don't mind at all, cause this will change as I said, do you want me to do that now?

Reply:
no it's fine, will see in the future